evidence/verify.go

```diff
@@ -25,26 +30,42 @@ func (evpool *Pool) verify(evidence types.Evidence) error {
 		ageNumBlocks   = height - evidence.Height()
 	)
 
-	// verify the time of the evidence
+	// ensure we have the block for the evidence height
+	//
+	// NOTE: It is currently possible for a peer to send us evidence we're not
+	// able to process because we're too far behind (e.g. syncing), so we DO NOT
+	// return an invalid evidence error because we do not want the peer to
+	// disconnect or signal an error in this particular case.
 	blockMeta := evpool.blockStore.LoadBlockMeta(evidence.Height())
 	if blockMeta == nil {
-		return fmt.Errorf("don't have header #%d", evidence.Height())
+		return fmt.Errorf("failed to verify evidence; missing block for height %d", evidence.Height())
 	}
+
+	// verify the time of the evidence
 	evTime := blockMeta.Header.Time
 	if evidence.Time() != evTime {
-		return fmt.Errorf("evidence has a different time to the block it is associated with (%v != %v)",
-			evidence.Time(), evTime)
+		return types.NewErrInvalidEvidence(
+			evidence,
+			fmt.Errorf(
+				"evidence has a different time to the block it is associated with (%v != %v)",
+				evidence.Time(), evTime,
+			),
+		)
 	}
+
 	ageDuration := state.LastBlockTime.Sub(evTime)
 
 	// check that the evidence hasn't expired
 	if ageDuration > evidenceParams.MaxAgeDuration && ageNumBlocks > evidenceParams.MaxAgeNumBlocks {
-		return fmt.Errorf(
-			"evidence from height %d (created at: %v) is too old; min height is %d and evidence can not be older than %v",
-			evidence.Height(),
-			evTime,
-			height-evidenceParams.MaxAgeNumBlocks,
-			state.LastBlockTime.Add(evidenceParams.MaxAgeDuration),
+		return types.NewErrInvalidEvidence(
+			evidence,
+			fmt.Errorf(
+				"evidence from height %d (created at: %v) is too old; min height is %d and evidence can not be older than %v",
+				evidence.Height(),
+				evTime,
+				height-evidenceParams.MaxAgeNumBlocks,
+				state.LastBlockTime.Add(evidenceParams.MaxAgeDuration),
+			),
```

Comment:
Just something to point out: You mentioned in a comment further up that we don't send invalid evidence when retrieving anything because we might be lagging behind the network and the evidence might be valid but we just don't have the blocks/validator set for it. In the opposite case where it is a peer that is lagging behind the node, the peer may send evidence that is still valid for them but no longer valid for the node in which case we will return an invalid evidence error. It's a pretty minor problem though.

Reply:
A bit hard to follow that one. Are you saying we _should not_ return a `ErrInvalidEvidence` in this exact case? Currently, the peer would most likely disconnect in the current context. Should we just return a normal error here?

Reply:
I think this is a niche case and that we still should return `ErrInvalidEvidence` in order to avoid peers from spamming the node with expired evidence